src/lib/device/journal_vdev.cpp

```diff
@@ -559,54 +574,108 @@ void JournalVirtualDev::Descriptor::truncate(off_t truncate_offset) {
+        auto* private_data = r_cast< JournalChunkPrivate* >(const_cast< uint8_t* >(chunk->user_private()));
+        LOGINFOMOD(journalvdev, "log_dev={} chunk_id={} is_head={}", m_logdev_id, chunk->chunk_id(),
+                   private_data->is_head)
+    }
+#endif
+
+    off_t cover_offset = sisl::round_down(data_start_offset(), chunk_size);
+    auto total_num_chunks = m_journal_chunks.size();
     for (auto it = m_journal_chunks.begin(); it != m_journal_chunks.end();) {
         auto chunk = *it;
-        start += chunk->size();
+        off_t end_of_chunk = m_vdev.get_end_of_chunk(chunk);
+        off_t chunk_hole = 0;
+        if (index == total_num_chunks - 1) {
+            // If its the last chunk, only read upto the tail_offset
+            // There are no holes in the last chunk.
+            cover_offset += (tail_off % chunk_size);
+        } else {
+            // For other chunks take the whole size.
+            cover_offset += chunk_size;
+            chunk_hole = (chunk_size - end_of_chunk);
+        }
 
-        // Also if its the last chunk and there is no data after truncate, we release chunk.
-        auto write_sz_in_total = m_write_sz_in_total.load(std::memory_order_relaxed);
-        if (start >= truncate_offset) { break; }
+        index++;
 
-        m_total_size -= chunk->size();
-        it = m_journal_chunks.erase(it);
+        // Check if the offset is inside the truncate offset. We also check if the truncate offset lies
+        // between end_of_chunk - chunk_size. If either condition satifies, we release the chunks.
+        if (cover_offset <= truncate_offset || ((cover_offset - chunk_hole) <= truncate_offset)) {
+            m_total_size -= chunk->size();
+            it = m_journal_chunks.erase(it);
 
-        // Clear the private chunk data before adding to pool.
-        auto* data = r_cast< JournalChunkPrivate* >(const_cast< uint8_t* >(chunk->user_private()));
-        *data = JournalChunkPrivate{};
-        m_vdev.update_chunk_private(chunk, data);
+            // Release the chunk back to pool.
+            LOGINFOMOD(journalvdev,
+                       "Released chunk_id={} log_dev={} cover={} truncate_offset={} tail={} end_of_chunk={} desc {}",
+                       chunk->chunk_id(), m_logdev_id, to_hex(cover_offset), to_hex(truncate_offset), to_hex(tail_off),
+                       m_vdev.get_end_of_chunk(chunk), to_string());
+            m_vdev.release_chunk_to_pool(chunk);
+        } else {
+            ++it;
+        }
+    }
 
-        // We ideally want to zero out chunks as chunks are reused after free across
-        // logdev's. But zero out chunk is very expensive, We look at crc mismatches
-        // to know the end offset of the log dev during recovery.
-        // Format and add back to pool.
-        m_vdev.m_chunk_pool->enqueue(chunk);
-        LOGINFOMOD(journalvdev, "After truncate released chunk {}", chunk->to_string());
+    if (update_truncate_offset) {
```

Comment:
is it expected data start offset won't be updated in the case all chunks can be released and `update_truncate_offset` left as false?

Reply:
Yeas as discussed above, its in memory. The actual data start offset is returned the caller. Which stores in metablk. LogDev::truncate

Reply:
see my concerns above. Not clear on the data start offset returned on else branch above.